pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/RealtimeTableDataManager.java

```diff
@@ -203,6 +207,36 @@ protected void doInit() {
       Preconditions.checkState(schema != null, "Failed to find schema for table: %s", _tableNameWithType);
       _tableUpsertMetadataManager = TableUpsertMetadataManagerFactory.create(tableConfig, schema, this, _serverMetrics);
     }
+
+    // For dedup and partial-upsert, need to wait for all segments loaded before starting consuming data
+    if (isDedupEnabled() || isPartialUpsertEnabled()) {
+      _isTableReadyToConsumeData = new BooleanSupplier() {
+        volatile boolean _allSegmentsLoaded = false;
+        long _lastCheckTimeNs = Long.MIN_VALUE;
+
+        @Override
+        public boolean getAsBoolean() {
+          if (_allSegmentsLoaded) {
+            return true;
+          } else {
+            synchronized (this) {
+              if (_allSegmentsLoaded) {
```

Comment:
Is the check down twice (first on line 219 and then on 223) to avoid taking unnecessary lock when the variable is true?

Reply:
Correct. We want to avoid lock when all segments are already loaded